command-not-found.sh

```diff
@@ -4,36 +4,80 @@
 # this will be called when a command is entered
 # but not found in the user’s path + environment
 command_not_found_handle () {
-    toplevel=nixos # TODO: detect this somehow
-    cmd=$1 # TODO: differentiate between paths and commands
+
+    # TODO: use "command not found" gettext translations
+
+    # taken from http://www.linuxjournal.com/content/bash-command-not-found
+    # - do not run when inside Midnight Commander or within a Pipe
+    if [ -n "$MC_SID" ] || ! [ -t 1 ]; then
+        >&2 echo "$1: command not found"
+        return 127
+    fi
+
+    toplevel=nixpkgs # nixpkgs should always be available even in NixOS
+    cmd=$1
     attrs=$(@out@/bin/nix-locate --minimal --no-group --type x --top-level --whole-name --at-root "/bin/$cmd")
     len=$(echo -n "$attrs" | grep -c "^")
 
     case $len in
         0)
             >&2 echo "$cmd: command not found"
             ;;
-        1) # TODO: add option to autorun with 1 match
-            >&2 echo "The program '$cmd' is currently not installed. You can install it"
-            >&2 echo "by typing:"
-            >&2 echo "  nix-env -iA $toplevel.$attrs"
+        1)
+            # if only 1 package provides this,
+            # then we can invoke it without asking the users
+            # of course if they have opted in with one of 2
+            # environment variables
+            # they are based on the ones in command-not-found.sh
+            #   NIX_AUTO_INSTALL : install the missing command into the
+            #                      user’s environment
+            #   NIX_AUTO_RUN     : run the command transparently inside of
+            #                      nix shell
+            # these will not return 127 if they worked correctly
+
+            # QUESTION: will this mess up some scripts if a broken script
+            #           fails to install?
```

Comment:
That would work fine for ```nix-env -iA```, but I'm not sure how to handle it with nix-shell. How can you differentiate between nix-shell installing something and running a command that fails and nix-shell just failing? Maybe we could run nix-build ahead of time to ensure that it builds correctly (although that seems hacky).